### ISSUE TITLE
Suggestion: Fixing/caching data to create consistent datasets

### ISSUE THREAD

**Issue by tminich** [FakerPHP/Faker#285]
Currently every call to a faker data generator function generates fresh random data. What I'd love to see is an option to switch faker into a cached mode where calls to the same function will generate data once and then always return the same data until caching is switched off.
Since Faker internally calls generator functions to create more complex data this would allow to create consistent datasets. To elaborate:
Lets say I want to create a set for a person, currently what I would get is something like this:
- First name: Thomas
- Last name: Miller
- Email: mary.rosen@example.com

With caching this would give me something like this:
- First name: Thomas
- Last name: Miller
- Email: t.miller@example.org

**Comment by prinx** [comment]
You can do so by `seed`ing the generator.

https://fakerphp.github.io/#seeding-the-generator

**Comment by tminich** [comment]
That unfortunately doesn't work. It guarantees a certain order of results but not the same result for let's say $faker->firstName every time it is called, especially internally. You can't even do RNG manipulation to get what you want because I still want a random dataset, so the seed for the generation needs to be random. Which means in my example I cannot know how many random calls I need to make to get the RNG to the state it would get the right 'answers' to generate the 'correct' email address since it is based on userName which uses a random schema of firstName and/or lastName in random order.

Since I posted I did look further through the code though, I will try if I can work out something like $faker->unique() is implemented, either with caching or internally reseeding for every call.

**Comment by prinx** [comment]
If you want to generate the same data in different part of the script, you need to call the `seed` every time you are generating the data.

```php
$faker = Factory::create();

$faker->seed(30);
echo $faker->firstName."\n";
echo $faker->lastName."\n";
echo $faker->email."\n\n";

$faker->seed(30);
echo $faker->firstName."\n"; // Will give the same first name as above
echo $faker->lastName."\n"; // Will give the same last name as above
echo $faker->email."\n\n"; // Will give the same email as above
```

Just make sure you generate the data in the same order and you don't call any other methods in the middle. Eg:
```php
$faker = Factory::create();

$faker->seed(30);
echo $faker->firstName."\n";
echo $faker->lastName."\n";
echo $faker->email."\n\n";

$faker->seed(30);
echo $faker->firstName."\n"; // Will give the same first name as above
echo $faker->lastName."\n"; // Will give the same last name as above
echo $faker->phoneNumber."\n";
echo $faker->email."\n\n"; // THE EMAIL WILL DIFER FROM THE ONE ABOVE BECAUSE IT IS NOT CALLED AT THE SAME POSITION AS FIRSTLY
```

If this doesn't work for you, you may be obliged to implement it yourself by maybe creating an array and saving the first time you generate and using the array the subsequent times.

**Comment by tminich** [comment]
> If this doesn't work for you, you may be obliged to implement it yourself by maybe creating an array and saving the first time you generate and using the array the subsequent times.

No, unfortunately it doesn't. Yes, it will generate the same data, but the email will be random and not generated from the first/lastName I requested before.

I've since created an implementation that is workable but pretty hacky. There are a bunch of caveats hidden in how Faker generates different datapoints unfortunately. It works for me but not necessarily for others.

TLDR, it's not as easy as I hoped to get this to work. I have something that does for me. Feel free to close if this isn't something that could be considered for the package itself.

**Comment by prinx** [comment]
You can share your implementation. It may help in designing the next version of the package, or help someone submit a PR.

**Comment by pimjansen** [comment]
@tminich can you please add a code sample of what you are trying and the use case where it does not work?

Please also include all relevant information as mention in a bug template

**Comment by tminich** [comment]
> 
> @tminich can you please add a code sample of what you are trying and the use case where it does not work?
> 
> Please also include all relevant information as mention in a bug template

Sure. I'm using Faker 1.13.0 with PHP 7.4 right now.

Here is what I want to achieve as demonstrated by a Laravel Tinker log using my implementation of caching:
```
>>> $faker->firstName
=> "Marion"
>>> $faker->lastName
=> "Ward"
>>> $faker->company
=> "Keeling-Zemlak"
>>> $faker->companyEmail2
=> "ward.marion@keelingzemlak.com"
>>>
```
(Note: `companyEmail2` is just a rewrite of the vanilla `companyEmail` depending on `company` instead of `lastName` down the line for the domain)

I hope it's obvious how this isn't possible with vanilla Faker for the reasons I stated above.

**Comment by pimjansen** [comment]
Then i do not really see the usecase. With the seeding you can retrieve exact the same data every single time:

```
<?php

declare(strict_types=1);

require 'vendor/autoload.php';

$faker = \Faker\Factory::create();

for ($i = 0; $i < 100; $i++) {
    $faker->seed(10);

    echo $faker->firstName.PHP_EOL;
    echo $faker->lastName.PHP_EOL;
    echo $faker->company.PHP_EOL;
    echo '--'.PHP_EOL;
}

/**
Ellis
Wyman
Nader-Maggio
--
Ellis
Wyman
Nader-Maggio
--
Ellis
Wyman
Nader-Maggio
--
 */
```

**Comment by tminich** [comment]
For these there data sure, but now include something that internally uses first/lastName and it won't work, sure you'll get the same value, but not based on 'Ellis' and 'Wyman'. In my example the email generated is based on the previously randomly generated values for firstName, LastName and company.

**Comment by pimjansen** [comment]
Ok this is something that faker will not do indeed since all are random elements. In this case you have to build them yourself
